### PR TITLE
Various improvements for kubeadm

### PR DESCRIPTION
**Pull request by luxas**
@errordeveloper

### CODE REVIEW

**Comment by errordeveloper** [comment]
@luxas thank you! ...you will need to rebase.

**Comment by lukemarsden** [comment]
These look like good changes, thank you @luxas! Have you built debs and tested them e2e with these code changes?

**Comment by luxas** [comment]
Nope, I haven't set up e2e yet

File: cmd/kubeadm/app/preflight/preflight.go
Comment (yoojinl) [inline cmd/kubeadm/app/preflight/preflight.go]: I have several comments:
1. It does not have to require root, permissions can be tuned in a way that other users can have permissions to perform required actions.
2. Requirements to have access to `kubelet` and `docker` commands breaks use-cases when kubeadm is run from the container. Or when `kubelet` is run in the container.
3. Requirement to have `docker` process running, may create problems for `rkt` use-cases.

We have next options:
1. Change those to just warning.
2. Add `--ignore-check`, so users have workaround.

As for me 1st option is more preferable.

Comment (luxas) [reply cmd/kubeadm/app/preflight/preflight.go]: In phase 1 (for the initial release) we do require root (discussed in the original kubeadm PR)
We will work on making it non-root later on

kubeadm will/should never be run in a container, it's a CLI tool.
kubelet will/should run in a container, but we'll remove the kubelet check when containerized kubelet is supported

the same for rkt, when we have rkt support, we'll check for both

Basically, this is just protecting users in the initial release
They should recieve helpful errors when they're doing something at-the-moment wrong

Comment (yoojinl) [reply cmd/kubeadm/app/preflight/preflight.go]: > In phase 1 (for the initial release) we do require root (discussed in the original kubeadm PR)

We don't require root, we require to have permissions for certain actions, and the user can provide such permissions for any registered user.

> kubeadm will/should never be run in a container, it's a CLI tool.

Many users prefer to have only docker and its dependencies run on the host, and run the rest in the containers, even tcpdump to debug something (and kubernetes community runs CLI tools in the container, e.g. to do builds).

> kubelet will/should run in a container, but we'll remove the kubelet check when containerized kubelet is supported

There are people who run kubelet in container, and it's up to the user to decide, if they want to run in the container or on host.

> Basically, this is just protecting users in the initial release
> They should recieve helpful errors when they're doing something at-the-moment wrong

I'm ok with providing the user helpful information, but I'm not sure if at the time we should prevent other users from using kubeadm, that is the reason it was suggested to provide a workaround `--ignore-checks`.

Comment (yoojinl) [reply cmd/kubeadm/app/preflight/preflight.go]: @errordeveloper @lukemarsden What do you think about that?

**Comment by lukemarsden** [comment]
sorry by e2e I just meant manual end-to-end testing @luxas

**Comment by luxas** [comment]
The only "breaking" change I can see here is the `address=localhost` thing but it _should really not_ impact. At least it works fine with a single-node

I've tested the CLI corner cases manually on my computer yes, and verified that k8s-version does work

@lukemarsden Is that enough?

File: cmd/kubeadm/app/preflight/preflight.go
Comment (errordeveloper) [inline cmd/kubeadm/app/preflight/preflight.go]: This can detect any short-lived `docker` process. It's a good start, but I am afraid it's going to be a bit fragile as is.

File: cmd/kubeadm/app/cmd/join.go
Comment (errordeveloper) [inline cmd/kubeadm/app/cmd/join.go]: I'd like this to be a separate command also. May be for the moment we should only provide this functionality as a separate command.

File: cmd/kubeadm/app/preflight/preflight.go
Comment (errordeveloper) [reply cmd/kubeadm/app/preflight/preflight.go]: Yes, containerised kubelet is one of the use-cases we want to support. Additionally, user might not have kubelet in the path, or anything like this may be possible.

**Review by errordeveloper — CHANGES_REQUESTED** [review]
This is an amazing start for pre-flight code, but I am afraid this is very last minute I don't have time to review or test it properly. However, I think we can merge this if you provide verification functionality as `kubeadm check`, so we can iterate on this without impacting the UX with many things that may turn out false negative.

File: cmd/kubeadm/app/preflight/preflight.go
Comment (errordeveloper) [inline cmd/kubeadm/app/preflight/preflight.go]: This isn't strictly necessary, write access to `/etc/kubernetes` is most that is required, and this can achieved without UID 0.

File: cmd/kubeadm/app/preflight/preflight.go
Comment (errordeveloper) [inline cmd/kubeadm/app/preflight/preflight.go]: I wouldn't rely simply on process name for anything whatsoever, especially a list of pids.

**Comment by errordeveloper** [comment]
@luxas I've cherry-picked the first commit, but renamed the flag to `--use-kubernetes-version`. If you get a chance to move the pre-flight check code into a separate command before Monday, it'd be great, but we could always add it in follow-up PR to master later on!

**Comment by luxas** [comment]
@errordeveloper Thanks for cherrypicking the first commit!
The verify stuff have to wait for now, but here are some other small fixes + plus added a version command.
That will surely prove to be valuable when users have problems!

PTAL

File: cmd/kubeadm/app/master/manifests.go
Comment (errordeveloper) [inline cmd/kubeadm/app/master/manifests.go:247]: I am not sure about adding this, per Mike's comment https://github.com/kubernetes/kubernetes/pull/33262#discussion_r80314282.

**Review by errordeveloper — COMMENTED** [review]
LGTM, but please clarify the need for `--address=127.0.0.1` or remove it.

**Comment by luxas** [comment]
I'm just like "Why expose ports on the master when we don't have to? Although the attack area is low"

@mikedanese

**Comment by luxas** [comment]
@errordeveloper Can we get this in now?
I seriously don't know why we should expose the ports although they handlers expose basically nothing.

File: cmd/kubeadm/app/kubeadm.go
Comment (lukemarsden) [inline cmd/kubeadm/app/kubeadm.go:50]: does this exist yet? we shouldn't make this change until it does

File: cmd/kubeadm/app/master/manifests.go
Comment (lukemarsden) [inline cmd/kubeadm/app/master/manifests.go:223]: Please can you explain what this change (and the `--address` changes below) does exactly?

File: cmd/kubeadm/app/master/manifests.go
Comment (lukemarsden) [reply cmd/kubeadm/app/master/manifests.go:247]: quoting Mike's comment here for context

> controller-manager and scheduler rest handlers expose basically nothing.

sounds like we can discuss adding this change later? I don't feel confident reviewing this change as-is, because I don't deeply understand the implications (and currently it's hard for me to test `kubeadm` changes, because it depends on a build env @errordeveloper has set up).

File: cmd/kubeadm/app/cmd/init.go
Comment (lukemarsden) [inline cmd/kubeadm/app/cmd/init.go]: Missing `\n` at the end of this Printf statement. Have you tested this manually and found that the code works?

File: cmd/kubeadm/app/api/types.go
Comment (lukemarsden) [inline cmd/kubeadm/app/api/types.go:91]: What is the motivation for this change below? Is it that it allows the user to have the message in https://github.com/errordeveloper/kubernetes/pull/17/files#r80388324 printed?

File: cmd/kubeadm/app/api/types.go
Comment (lukemarsden) [inline cmd/kubeadm/app/api/types.go]: What's the motivation for moving this to the other module?

File: cmd/kubeadm/app/api/types.go
Comment (lukemarsden) [inline cmd/kubeadm/app/api/types.go]: Can you describe exactly where the v0.1.0 `kubeadm` version number will appear versus the `1.5.0-alpha` that will appear in the package version for `kubeadm`? I'm concerned that showing two different version numbers will be potentially confusing for users. I'd prefer to help users differentiate between kubeadm releases with date- or sha1-based versioning for now, at least until we get a chance to discuss this in the SIG.

**Comment by lukemarsden** [comment]
Thank you for the contribution @luxas! I am wary of merging this as-is because as far as I know no one has manually tested the code changes yet. I'm also not convinced that the code changes are required before we lgtm https://github.com/kubernetes/kubernetes/pull/33262. As far as I can tell the code changes here are:
- `kubeadm version`, while I see the potential value to users, I think this merits further discussion about the `0.1.0` vs `1.5.0-alpha` version strings that will be visible to users, how they interact, and how we should version `kubeadm` in the interval between 1.4 and 1.5. IMO this should not block #33262 (but you could make the `version` PR against `kubeadm` as soon as it's in kubernetes master and we could have the discussion there 😄).
- a minor change to UX when user has provided `--cloud-provider` argument (cloud provider isn't being documented in the [kubeadm docs yet](https://github.com/kubernetes/kubernetes.github.io/pull/1265), so IMO this is lower priority). I don't have an easy way to test this change so it's hard to verify it.
- pointing to where the discovery image will be (but isn't yet). Until that image is published, this change actually makes things worse not better, since if we merge this new builds of `kubeadm` will break.

To summarize, I think we should lgtm https://github.com/kubernetes/kubernetes/pull/33262 as-is and then make the above changes as separate PRs against kubernetes master, which can then be discussed and merged separately. How does that sound @luxas?

File: cmd/kubeadm/app/kubeadm.go
Comment (luxas) [reply cmd/kubeadm/app/kubeadm.go:50]: It exists.
I wouldn't do the change if it didn't

File: cmd/kubeadm/app/master/manifests.go
Comment (luxas) [reply cmd/kubeadm/app/master/manifests.go:223]: address is deprecated.
insecure-bind-address is the new equalient flag

Comment (luxas) [reply cmd/kubeadm/app/master/manifests.go:223]: Nothing changes

File: cmd/kubeadm/app/api/types.go
Comment (luxas) [reply cmd/kubeadm/app/api/types.go]: It makes more sense there plus being able to output it for the user if there's need to

File: cmd/kubeadm/app/master/manifests.go
Comment (lukemarsden) [reply cmd/kubeadm/app/master/manifests.go:223]: ahh, thanks

File: cmd/kubeadm/app/kubeadm.go
Comment (lukemarsden) [reply cmd/kubeadm/app/kubeadm.go:50]: oh, yay!

**Comment by lukemarsden** [comment]
Sorry, I was wrong about the gcr.io image. @errordeveloper let's test this PR together tomorrow and if it passes all my tests we can get this in, modulo the changes that @luxas is about to make :)

**Comment by lukemarsden** [comment]
Latest changes LGTM from my reading of the code @luxas. I'll test this tomorrow with @errordeveloper and merge this as soon as we're happy with it. Thanks!

**Comment by luxas** [comment]
Great! I think we should be in good shape after this one

**Comment by lukemarsden** [comment]
Awesome ✨

**Comment by lukemarsden** [comment]
LGTM, merging, @luxas says he has manually tested it. Will summarize changeset in parent PR momentarily.

**Comment by lukemarsden** [comment]
Changeset description: https://github.com/kubernetes/kubernetes/pull/33262#issuecomment-249442470

**Pull request merged**
